cupy/core/_routines_math.pyx

```diff
@@ -253,6 +267,24 @@ _sum_keep_dtype = create_reduction_func(
     ('in0', 'a + b', 'out0 = type_out0_raw(a)', None), 0)
 
 
+_nansum_auto_dtype = create_reduction_func(
+    'cupy_nansum',
+    ('?->l', 'b->l', 'B->L', 'h->l', 'H->L', 'i->l', 'I->L', 'l->l', 'L->L',
+     'q->q', 'Q->Q',
+     ('e->e', (None, None, None, 'float')),
+     'f->f', 'd->d', 'F->F', 'D->D'),
+    ('in0', '(a == a) * a + (b == b) * b', 'out0 = type_out0_raw(a)', None), 0)
```

Comment:
It seems to check if the value is NaN in each binary operation. How about the following implementation?
```py
_nansum_auto_dtype = create_reduction_func(
    'cupy_nansum',
    ('?->l', 'b->l', 'B->L', 'h->l', 'H->L', 'i->l', 'I->L', 'l->l', 'L->L',
     'q->q', 'Q->Q',
     ('e->e', (None, None, None, 'float')),
     'f->f', 'd->d', 'F->F', 'D->D'),
    ('(in0 == in0) * in0', 'a + b', 'out0 = type_out0_raw(a)', None), 0)
```

Reply:
I didn't know this was allowed, thanks for the great suggestion! Done.